### ISSUE TITLE
rc4_encrypt crash

### ISSUE THREAD

**Issue by damolp** [arvidn/libtorrent#4074]
**Please provide the following information**

libtorrent version (or branch):
1.2.0

platform/architecture:
Ubuntu 18.04.3 LTS x86_64

compiler and compiler version:
gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.
I'm seeing crashes on the rc4_encrypt function within pe_crypto.cpp. The out address is invalid so I think we need a check on [this cast?](https://github.com/arvidn/libtorrent/blob/d33b0506a0f4cd61860fd54864267c82d3e171e3/src/pe_crypto.cpp#L317)

Reproducing is hard, I only see it once in a blue moon. I'll try fuzzing the RC4 functions to see if I can find input that will reproduce it.

```
#0  0x00007fdcf0298229 in libtorrent::rc4_encrypt (out=0x1 <error: Cannot access memory at address 0x1>, outlen=16383, state=state@entry=0x7fd35fbbbba0) at pe_crypto.cpp:406
#1  0x00007fdcf0298328 in libtorrent::rc4_handler::encrypt (this=0x7fd35fbbba90, bufs=...) at pe_crypto.cpp:324
#2  0x00007fdcf0297ae4 in libtorrent::encryption_handler::encrypt (this=this@entry=0x7fd322866730, iovec=...) at pe_crypto.cpp:168
#3  0x00007fdcf01faa21 in libtorrent::bt_peer_connection::hit_send_barrier (this=0x7fd322865ff0, iovec=...) at bt_peer_connection.cpp:3343
#4  0x00007fdcf02b5a9d in libtorrent::peer_connection::setup_send (this=this@entry=0x7fd322865ff0) at peer_connection.cpp:5517
#5  0x00007fdcf02b651b in libtorrent::peer_connection::on_send_data (this=0x7fd322865ff0, error=..., bytes_transferred=13) at peer_connection.cpp:6230


#0  0x00007fd729585229 in libtorrent::rc4_encrypt (out=0x1 <error: Cannot access memory at address 0x1>, outlen=16383, state=state@entry=0x7fd71fe525d0) at pe_crypto.cpp:406
#1  0x00007fd729585328 in libtorrent::rc4_handler::encrypt (this=0x7fd71fe524c0, bufs=...) at pe_crypto.cpp:324
#2  0x00007fd729584ae4 in libtorrent::encryption_handler::encrypt (this=this@entry=0x7fd3642531e0, iovec=...) at pe_crypto.cpp:168
#3  0x00007fd7294e7a21 in libtorrent::bt_peer_connection::hit_send_barrier (this=0x7fd364252aa0, iovec=...) at bt_peer_connection.cpp:3343
#4  0x00007fd7295a2a9d in libtorrent::peer_connection::setup_send (this=this@entry=0x7fd364252aa0) at peer_connection.cpp:5517
#5  0x00007fd7295a351b in libtorrent::peer_connection::on_send_data (this=0x7fd364252aa0, error=..., bytes_transferred=13) at peer_connection.cpp:6230

```

**Comment by arvidn** [comment]
if you implement a fuzzer, please consider making a pull request! there is a collections of fuzzers in libtorrent already

**Comment by arvidn** [comment]
as far as I can tell from your stack traces, the length of the buffer is still 16383, so the pointer is supposed to be valid. checking for null would not solve this problem most likely, as some corruption would have already been introduced somewhere

**Comment by arvidn** [comment]
it might be a bit far-fetched; but any chance this patch fixes this? https://github.com/arvidn/libtorrent/pull/4157

**Comment by damolp** [comment]
Will test it this weekend, cheers for looking into it.

**Comment by arvidn** [comment]
This was released in 1.2.3. Any chance this fixed the crash?

**Comment by damolp** [comment]
Yeah looks like it has fixed it. I'm getting a few different crashes now though on 1.2.3 (I'll open a new issue once I get core dumps and can pinpoint where they are)

```
[222865.675278] traps: deluged[77897] general protection ip:7fccc7417208 sp:7fcc92181e88 error:0 in libtorrent-rasterbar.so.10.0.0[7fccc7357000+3d8000]
[493269.208711] deluged[82333]: segfault at 56c ip 00007fdd54688fc9 sp 00007fdd5255bba0 error 4 in libtorrent-rasterbar.so.10.0.0[7fdd54433000+3d8000]
[615721.739432] traps: deluged[42252] general protection ip:7f2f65531788 sp:7ffdc3efcd00 error:0 in libtorrent.cpython-36m-x86_64-linux-gnu.so[7f2f6549f000+19e000]
[728846.580113] traps: python3[9763] general protection ip:7f49460c4788 sp:7ffdffb401b0 error:0 in libtorrent.cpython-36m-x86_64-linux-gnu.so[7f4946032000+19e000]
```
